File: android_adt/src/org/tvp/kirikiri2/KR2Activity.java
package org.tvp.kirikiri2;

import android.app.Activity;
import android.app.ActivityManager;
import android.app.AlertDialog;
import android.content.Context;
import android.content.DialogInterface;
import android.content.pm.PackageInfo;
import android.content.pm.PackageManager;
import android.os.Bundle;
import android.os.Handler;
import android.os.Looper;
import android.provider.Settings.Secure;
import android.util.Log;
import android.view.View;
import android.view.ViewGroup;
import android.widget.EditText;
import android.widget.FrameLayout;

import java.io.File;
import java.io.FileOutputStream;
import java.io.IOException;
import java.util.ArrayList;
import java.util.List;
import java.util.Locale;

//import android.os.Build.VERSION;
public class KR2Activity extends Activity {
   public static final Companion Companion = new Companion();
   private static final String TAG = "krkr2";
   private static volatile KR2Activity instance;
   private static volatile Context appContext;
   private static final Handler uiHandler = new Handler(Looper.getMainLooper());

   @Override
   protected void onCreate(Bundle savedInstanceState) {
      super.onCreate(savedInstanceState);
      instance = this;
      appContext = this.getApplicationContext();
   }

   @Override
   protected void onDestroy() {
      if (instance == this) {
         instance = null;
      }
      super.onDestroy();
   }

   public static final KR2Activity GetInstance() {
      return Companion.GetInstance();
   }

   public static final void updateMemoryInfo() {
      Companion.updateMemoryInfo();
   }

   public static final long getAvailMemory() {
      return Companion.getAvailMemory();
   }

   public static final long getUsedMemory() {
      return Companion.getUsedMemory();
   }

   public static final String getDeviceId() {
      return Companion.getDeviceId();
   }

   public static final String GetVersion() {
      return Companion.GetVersion();
   }

   public static final String getLocaleName() {
      return Companion.getLocaleName();
   }

   public static final String[] getStoragePath() {
      return Companion.getStoragePath();
   }

   public static final boolean isWritableNormalOrSaf(String path) {
      return Companion.isWritableNormalOrSaf(path);
   }

   public static final boolean CreateFolders(String path) {
      return Companion.CreateFolders(path);
   }

   public static final boolean WriteFile(String path, byte[] data) {
      return Companion.WriteFile(path, data);
   }

   public static final boolean DeleteFile(String path) {
      return Companion.DeleteFile(path);
   }

   public static final boolean RenameFile(String from, String to) {
      return Companion.RenameFile(from, to);
   }

   public static final void MessageController(int adType, int arg1, int arg2) {
      Companion.MessageController(adType, arg1, arg2);
   }

   public static final void ShowMessageBox(String title, String message, String[] buttons) {
      Companion.ShowMessageBox(title, message, buttons);
   }

   public static final void ShowInputBox(String title, String prompt, String initText, String[] buttons) {
      Companion.ShowInputBox(title, prompt, initText, buttons);
   }

   public static final void hideTextInput() {
      Companion.hideTextInput();
   }

   public static final void showTextInput(int x, int y, int w, int h) {
      Companion.showTextInput(x, y, w, h);
   }

   public static final void exit() {
      Companion.exit();
   }

   private static final native void nativeOnMessageBoxResult(int var0);

   private static final native void nativeOnInputBoxResult(int var0, String var1);

   public static final class Companion {
      private Companion() {
      }

      public final KR2Activity GetInstance() {
         return KR2Activity.instance;
      }

      public final void updateMemoryInfo() {
      }

      public final long getAvailMemory() {
         if (KR2Activity.appContext == null) {
            return 0L;
         } else {
            Context ctx = KR2Activity.appContext;
            Object var3 = ctx.getSystemService("activity");
            ActivityManager am = var3 instanceof ActivityManager ? (ActivityManager)var3 : null;
            ActivityManager.MemoryInfo info = new ActivityManager.MemoryInfo();
            if (am != null) {
               am.getMemoryInfo(info);
            }
            return info.availMem;
         }
      }

      public final long getUsedMemory() {
         if (KR2Activity.appContext == null) {
            return 0L;
         } else {
            Context ctx = KR2Activity.appContext;
            Object var3 = ctx.getSystemService("activity");
            ActivityManager am = var3 instanceof ActivityManager ? (ActivityManager)var3 : null;
            ActivityManager.MemoryInfo info = new ActivityManager.MemoryInfo();
            if (am != null) {
               am.getMemoryInfo(info);
            }
            return info.totalMem - info.availMem;
         }
      }

      public final String getDeviceId() {
         if (KR2Activity.appContext == null) {
            return "";
         } else {
            Context ctx = KR2Activity.appContext;
            String var2 = Secure.getString(ctx.getContentResolver(), "android_id");
            if (var2 == null) {
               var2 = "";
            }
            return var2;
         }
      }


      public final String GetVersion() {
         if (KR2Activity.appContext == null) {
            return "";
         } else {
            Context ctx = KR2Activity.appContext;

            try {
               PackageManager pm = ctx.getPackageManager();
               String pkg = ctx.getPackageName();
               PackageInfo pi = /*VERSION.SDK_INT >= 33 ? pm.getPackageInfo(pkg, PackageInfoFlags.of(0L)) :*/ pm.getPackageInfo(pkg, 0);
               String var7 = pi.versionName;
               if (var7 == null) {
                  var7 = "";
               }
               return var7;
            } catch (Exception var5) {
               return "";
            }
         }
      }

      public final String getLocaleName() {
         //Intrinsics.checkNotNullExpressionValue(var10000, "getLanguage(...)");
         return Locale.getDefault().getLanguage();
      }

      public final String[] getStoragePath() {
         if (KR2Activity.appContext == null) {
            return new String[0];
         } else {
            Context ctx = KR2Activity.appContext;
            List<String> paths = new ArrayList<String>();
            File[] var13 = ctx.getExternalFilesDirs((String)null);
            if (var13 != null) {
               int var6 = 0;
               for(int var7 = var13.length; var6 < var7; ++var6) {
                  Object element$iv = var13[var6];
                  if (element$iv != null) {
                     //Intrinsics.checkNotNullExpressionValue(var10001, "getAbsolutePath(...)");
                     paths.add(((File)element$iv).getAbsolutePath());
                  }
               }
            }

            File internal = ctx.getFilesDir();
            if (internal != null) {
               String var14 = internal.getAbsolutePath();
               //Intrinsics.checkNotNullExpressionValue(var14, "getAbsolutePath(...)");
               paths.add(var14);
            }

            return paths.toArray(new String[0]);
         }
      }

      public final boolean isWritableNormalOrSaf(String path) {
         //Intrinsics.checkNotNullParameter(path, "path");
         try {
            File file = new File(path);
            if (!file.exists()) {
               file.mkdirs();
            }
            return file.canWrite();
         } catch (Exception var4) {
            return false;
         }
      }

      public final boolean CreateFolders(String path) {
         //Intrinsics.checkNotNullParameter(path, "path");
         try {
            return (new File(path)).mkdirs();
         } catch (Exception var4) {
            return false;
         }
      }

      public final boolean WriteFile(String path, byte[] data) {
         //Intrinsics.checkNotNullParameter(path, "path");
         //Intrinsics.checkNotNullParameter(data, "data");

         try {
            //FilesKt.writeBytes(new File(path), data);
        	 FileOutputStream it = null;
        	 try {
        		 it = new FileOutputStream(new File(path));
        		 it.write(data);
        	 } catch (IOException e) {
        		 e.printStackTrace();
        	 } finally {
        		 if (it != null) {
        			 it.close();
        		 }
        	 }
        	 return true;
         } catch (Exception var5) {
        	 return false;
         }
      }

      public final boolean DeleteFile(String path) {
         //Intrinsics.checkNotNullParameter(path, "path");
    	 try {
            return (new File(path)).delete();
         } catch (Exception var4) {
            return false;
         }
      }

      public final boolean RenameFile(String from, String to) {
         //Intrinsics.checkNotNullParameter(from, "from");
         //Intrinsics.checkNotNullParameter(to, "to");
         try {
            return (new File(from)).renameTo(new File(to));
         } catch (Exception var5) {
            return false;
         }
      }

      public final void MessageController(int adType, int arg1, int arg2) {
         Log.d("krkr2", "MessageController ignored: " + adType + ' ' + arg1 + ' ' + arg2);
      }

      public final void ShowMessageBox(final String title, final String message, final String[] buttons) {
//         Intrinsics.checkNotNullParameter(title, "title");
//         Intrinsics.checkNotNullParameter(message, "message");
//         Intrinsics.checkNotNullParameter(buttons, "buttons");
         KR2Activity var10000 = KR2Activity.instance;
         if (var10000 == null) {
            Log.w("krkr2", "ShowMessageBox: no activity, auto-confirming");
            this.nativeOnMessageBoxResult(0);
         } else {
            final KR2Activity ctx = var10000;
            KR2Activity.uiHandler.post(new Runnable() {
				@Override
				public void run() {
			         try {
			             AlertDialog.Builder var10000 = (new AlertDialog.Builder((Context)ctx)).setTitle((CharSequence)title).setMessage((CharSequence)message).setCancelable(false);
			             //Intrinsics.checkNotNullExpressionValue(var10000, "setCancelable(...)");
			             AlertDialog.Builder builder = var10000;
			             if (buttons.length != 0) {
			                builder.setPositiveButton((CharSequence)buttons[0], new DialogInterface.OnClickListener() {
								@Override
								public void onClick(DialogInterface arg0, int arg1) {
									KR2Activity.Companion.nativeOnMessageBoxResult(0);
								}
							});
			             }
			             if (buttons.length >= 2) {
			                builder.setNeutralButton((CharSequence)buttons[1], new DialogInterface.OnClickListener() {
								@Override
								public void onClick(DialogInterface arg0, int arg1) {
									KR2Activity.Companion.nativeOnMessageBoxResult(1);
								}
							});
			             }
			             if (buttons.length >= 3) {
			                builder.setNegativeButton((CharSequence)buttons[2], new DialogInterface.OnClickListener() {
								@Override
								public void onClick(DialogInterface arg0, int arg1) {
									KR2Activity.Companion.nativeOnMessageBoxResult(2);
								}
							});
			             }
			             builder.show();
			          } catch (Exception e) {
			             Log.e("krkr2", "ShowMessageBox failed", (Throwable)e);
			             KR2Activity.Companion.nativeOnMessageBoxResult(0);
			          }
				}
            });
         }
      }

      public final void ShowInputBox(final String title, final String prompt, final String initText, final String[] buttons) {
//         Intrinsics.checkNotNullParameter(title, "title");
//         Intrinsics.checkNotNullParameter(prompt, "prompt");
//         Intrinsics.checkNotNullParameter(initText, "initText");
//         Intrinsics.checkNotNullParameter(buttons, "buttons");
         if (KR2Activity.instance == null) {
            Log.w("krkr2", "ShowInputBox: no activity, auto-confirming");
            this.nativeOnInputBoxResult(0, initText);
         } else {
            final KR2Activity ctx = KR2Activity.instance;
            KR2Activity.uiHandler.post(new Runnable() {
				@Override
				public void run() {
			         try {
			             int dp24 = (int)((float)24 * ctx.getResources().getDisplayMetrics().density);
			             FrameLayout container = new FrameLayout((Context)ctx);
			             container.setPadding(dp24, 0, dp24, 0);
			             final EditText editText = new EditText((Context)ctx);
			             editText.setText((CharSequence)initText);
			             container.addView((View)editText, (ViewGroup.LayoutParams)(new FrameLayout.LayoutParams(-1, -2)));
			             AlertDialog.Builder builder = new AlertDialog.Builder((Context)ctx)
			            		 .setTitle((CharSequence)title)
			            		 .setMessage((CharSequence)prompt)
			            		 .setView((View)container)
			            		 .setCancelable(false);
			             //Intrinsics.checkNotNullExpressionValue(var10000, "setCancelable(...)");
			             if (buttons.length != 0) {
			                builder.setPositiveButton((CharSequence)buttons[0], new DialogInterface.OnClickListener() {
			                    @Override
			                    public void onClick(DialogInterface dialog, int which) {
			                    	KR2Activity.Companion.nativeOnInputBoxResult(0, editText.getText().toString());
			                    }
			                });
			             }

			             if (buttons.length >= 2) {
			                builder.setNeutralButton((CharSequence)buttons[1], new DialogInterface.OnClickListener() {
			                    @Override
			                    public void onClick(DialogInterface dialog, int which) {
			                    	KR2Activity.Companion.nativeOnInputBoxResult(1, editText.getText().toString());
			                    }
			                });
			             }

			             if (buttons.length >= 3) {
			                builder.setNegativeButton((CharSequence)buttons[2], new DialogInterface.OnClickListener() {
			                    @Override
			                    public void onClick(DialogInterface dialog, int which) {
			                    	KR2Activity.Companion.nativeOnInputBoxResult(2, editText.getText().toString());
			                    }
			                });
			             }

			             builder.show();
			             editText.requestFocus();
			          } catch (Exception e) {
			             Log.e("krkr2", "ShowInputBox failed", (Throwable)e);
			             KR2Activity.Companion.nativeOnInputBoxResult(0, initText);
			          }
				}
            });
         }
      }

      public final void hideTextInput() {
      }

      public final void showTextInput(int x, int y, int w, int h) {
      }

      public final void exit() {
         if (KR2Activity.instance != null) {
        	 KR2Activity.instance.finish();
         }
      }

      private final void nativeOnMessageBoxResult(int result) {
         KR2Activity.nativeOnMessageBoxResult(result);
      }

      private final void nativeOnInputBoxResult(int result, String text) {
         KR2Activity.nativeOnInputBoxResult(result, text);
      }
   }
}

File: apps/flutter_app/android/app/src/main/kotlin/org/tvp/kirikiri2/KR2Activity.kt
package org.tvp.kirikiri2

import android.app.ActivityManager
import android.content.Context
import android.os.Build
import android.os.Bundle
import android.os.Handler
import android.os.Looper
import android.provider.Settings
import android.util.Log
import android.widget.EditText
import android.widget.FrameLayout
import com.google.android.material.dialog.MaterialAlertDialogBuilder
import io.flutter.embedding.android.FlutterActivity
import java.io.File
import java.util.Locale

open class KR2Activity : FlutterActivity() {
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        instance = this
        appContext = applicationContext
    }

    override fun onDestroy() {
        if (instance === this) {
            instance = null
        }
        super.onDestroy()
    }

    companion object {
        private const val TAG = "krkr2"

        @Volatile
        private var instance: KR2Activity? = null

        @Volatile
        private var appContext: Context? = null

        @JvmStatic
        fun GetInstance(): KR2Activity? {
            return instance
        }

        @JvmStatic
        fun updateMemoryInfo() {
            // No-op; memory info is fetched on demand via ActivityManager.
        }

        @JvmStatic
        fun getAvailMemory(): Long {
            val ctx = appContext ?: return 0L
            val am = ctx.getSystemService(Context.ACTIVITY_SERVICE) as? ActivityManager
            val info = ActivityManager.MemoryInfo()
            am?.getMemoryInfo(info)
            return info.availMem
        }

        @JvmStatic
        fun getUsedMemory(): Long {
            val ctx = appContext ?: return 0L
            val am = ctx.getSystemService(Context.ACTIVITY_SERVICE) as? ActivityManager
            val info = ActivityManager.MemoryInfo()
            am?.getMemoryInfo(info)
            return info.totalMem - info.availMem
        }

        @JvmStatic
        fun getDeviceId(): String {
            val ctx = appContext ?: return ""
            return Settings.Secure.getString(ctx.contentResolver, Settings.Secure.ANDROID_ID) ?: ""
        }

        @JvmStatic
        fun GetVersion(): String {
            val ctx = appContext ?: return ""
            return try {
                val pm = ctx.packageManager
                val pkg = ctx.packageName
                val pi = if (Build.VERSION.SDK_INT >= 33) {
                    pm.getPackageInfo(pkg, android.content.pm.PackageManager.PackageInfoFlags.of(0))
                } else {
                    @Suppress("DEPRECATION")
                    pm.getPackageInfo(pkg, 0)
                }
                pi.versionName ?: ""
            } catch (e: Exception) {
                ""
            }
        }

        @JvmStatic
        fun getLocaleName(): String {
            return Locale.getDefault().language
        }

        @JvmStatic
        fun getStoragePath(): Array<String> {
            val ctx = appContext ?: return emptyArray()
            val paths = mutableListOf<String>()
            ctx.getExternalFilesDirs(null)?.forEach { file ->
                if (file != null) paths.add(file.absolutePath)
            }
            val internal = ctx.filesDir
            if (internal != null) paths.add(internal.absolutePath)
            return paths.toTypedArray()
        }

        @JvmStatic
        fun isWritableNormalOrSaf(path: String): Boolean {
            return try {
                val file = File(path)
                if (!file.exists()) {
                    file.mkdirs()
                }
                file.canWrite()
            } catch (e: Exception) {
                false
            }
        }

        @JvmStatic
        fun CreateFolders(path: String): Boolean {
            return try {
                File(path).mkdirs()
            } catch (e: Exception) {
                false
            }
        }

        @JvmStatic
        fun WriteFile(path: String, data: ByteArray): Boolean {
            return try {
                File(path).writeBytes(data)
                true
            } catch (e: Exception) {
                false
            }
        }

        @JvmStatic
        fun DeleteFile(path: String): Boolean {
            return try {
                File(path).delete()
            } catch (e: Exception) {
                false
            }
        }

        @JvmStatic
        fun RenameFile(from: String, to: String): Boolean {
            return try {
                File(from).renameTo(File(to))
            } catch (e: Exception) {
                false
            }
        }

        @JvmStatic
        fun MessageController(adType: Int, arg1: Int, arg2: Int) {
            // Ads not supported in Flutter build.
            Log.d(TAG, "MessageController ignored: $adType $arg1 $arg2")
        }

        private val uiHandler = Handler(Looper.getMainLooper())

        @JvmStatic
        fun ShowMessageBox(title: String, message: String, buttons: Array<String>) {
            val ctx = instance ?: run {
                Log.w(TAG, "ShowMessageBox: no activity, auto-confirming")
                nativeOnMessageBoxResult(0)
                return
            }
            uiHandler.post {
                try {
                    val builder = MaterialAlertDialogBuilder(ctx)
                        .setTitle(title)
                        .setMessage(message)
                        .setCancelable(false)
                    if (buttons.isNotEmpty()) {
                        builder.setPositiveButton(buttons[0]) { _, _ -> nativeOnMessageBoxResult(0) }
                    }
                    if (buttons.size >= 2) {
                        builder.setNeutralButton(buttons[1]) { _, _ -> nativeOnMessageBoxResult(1) }
                    }
                    if (buttons.size >= 3) {
                        builder.setNegativeButton(buttons[2]) { _, _ -> nativeOnMessageBoxResult(2) }
                    }
                    builder.show()
                } catch (e: Exception) {
                    Log.e(TAG, "ShowMessageBox failed", e)
                    nativeOnMessageBoxResult(0)
                }
            }
        }

        @JvmStatic
        fun ShowInputBox(title: String, prompt: String, initText: String, buttons: Array<String>) {
            val ctx = instance ?: run {
                Log.w(TAG, "ShowInputBox: no activity, auto-confirming")
                nativeOnInputBoxResult(0, initText)
                return
            }
            uiHandler.post {
                try {
                    val dp24 = (24 * ctx.resources.displayMetrics.density).toInt()
                    val container = FrameLayout(ctx).apply {
                        setPadding(dp24, 0, dp24, 0)
                    }
                    val editText = EditText(ctx).apply {
                        setText(initText)
                    }
                    container.addView(editText, FrameLayout.LayoutParams(
                        FrameLayout.LayoutParams.MATCH_PARENT,
                        FrameLayout.LayoutParams.WRAP_CONTENT
                    ))
                    val builder = MaterialAlertDialogBuilder(ctx)
                        .setTitle(title)
                        .setMessage(prompt)
                        .setView(container)
                        .setCancelable(false)
                    if (buttons.isNotEmpty()) {
                        builder.setPositiveButton(buttons[0]) { _, _ ->
                            nativeOnInputBoxResult(0, editText.text.toString())
                        }
                    }
                    if (buttons.size >= 2) {
                        builder.setNeutralButton(buttons[1]) { _, _ ->
                            nativeOnInputBoxResult(1, editText.text.toString())
                        }
                    }
                    if (buttons.size >= 3) {
                        builder.setNegativeButton(buttons[2]) { _, _ ->
                            nativeOnInputBoxResult(2, editText.text.toString())
                        }
                    }
                    builder.show()
                    editText.requestFocus()
                } catch (e: Exception) {
                    Log.e(TAG, "ShowInputBox failed", e)
                    nativeOnInputBoxResult(0, initText)
                }
            }
        }

        @JvmStatic
        fun hideTextInput() {
            // No-op; Flutter manages IME.
        }

        @JvmStatic
        fun showTextInput(x: Int, y: Int, w: Int, h: Int) {
            // No-op; Flutter manages IME.
        }

        @JvmStatic
        fun exit() {
            instance?.finish()
        }

        @JvmStatic
        private external fun nativeOnMessageBoxResult(result: Int)

        @JvmStatic
        private external fun nativeOnInputBoxResult(result: Int, text: String)
    }
}
